microsoft/testsuites/dpdk/common.py

```diff
@@ -17,6 +22,261 @@
+        )
+
+
+class Downloader:
+    def __init__(self, node: Node) -> None:
+        self._node = node
+
+    def download(self) -> PurePath:
+        raise NotImplementedError("Downloader not implemented.")
+
+
+class GitDownload(Downloader):
+    git_repo: str = ""
+    git_ref: str = ""
+
+    def __init__(
+        self,
+        node: Node,
+        git_repo: str,
+        git_ref: str,
+    ) -> None:
+        super().__init__(node)
+        self.git_repo = git_repo
+        self.git_ref = git_ref
+
+    # checkout the git repository into the working path
+    def download(self) -> PurePath:
+        # NOTE: fail on exists is set to True.
+        # The expectation is that the parent Installer class should
+        # remove any lingering installations
+        self._source_path = self._node.tools[Git].clone(
+            self.git_repo,
+            cwd=self._node.get_working_path(),
+            ref=self.git_ref,
+            fail_on_exists=True,
+        )
+        self._branch_handler()
+        return self._source_path
+
+    # allow children to apply an optional
+    # post-checkout branch handler
+    def _branch_handler(self) -> None:
+        ...
+
+
+# parent class for tarball source installations
+class TarDownload(Downloader):
+    def __init__(
+        self,
+        node: Node,
+        tar_url: str,
+    ) -> None:
+        super().__init__(node)
+        self._tar_url = tar_url
+        self._is_remote_tarball = tar_url.startswith("https://")
+
+    # fetch the tarball (or copy it to the node)
+    # then extract it
+    def download(self) -> PurePath:
+        node = self._node
+        work_path = self._node.get_working_path()
+        if self._is_remote_tarball:
+            tarfile = node.tools[Wget].get(
+                self._tar_url,
+                file_path=str(work_path),
+                overwrite=False,
+                force_run=True,
+            )
+            remote_path = node.get_pure_path(tarfile)
+            self.tar_filename = remote_path.name
+        else:
+            self.tar_filename = PurePath(self._tar_url).name
+            remote_path = work_path.joinpath(self.tar_filename)
+            node.shell.copy(
+                local_path=PurePath(self._tar_url),
+                node_path=remote_path,
+            )
+        # get rid of suffixes like .tar[.gz]
+        self._source_path = work_path
+        for suffix in [".tar.gz", ".tar.bz2", ".tar"]:
+            if self.tar_filename.endswith(suffix):
+                unpacked_filename = self.tar_filename[: -len(suffix)]
+                self._source_path = work_path.joinpath(unpacked_filename)
+                break
+        assert_that(str(self._source_path)).described_as(
+            "Source path was not set by tar installer, "
+            "likely an unexpected file extension for "
+            f"the tar file: {self.tar_filename}"
+        ).is_not_equal_to(str(work_path))
```

Comment:
The logic is not easy to understand. It should assert there is an tar file, and unpack it. It shouldn't do above steps silently, and then check if the path have tar or not.

Reply:
I'll fix.